File: 2022/solutions/02-solution.R
setwd("/Users/tea/coding/advent-of-code-2022")

##### 02 #####
### Part 1
input <- read.table("input/02-input.txt")

rps <- function(input){
    points <- 0
    for (i in 1:nrow(input)) {
       if (input[i, 1] == "A" && input[i, 2] == "X") {
            points <- points + 3
       }
       if (input[i, 1] == "B" && input[i, 2] == "Y") {
            points <- points + 3
       }
       if (input[i, 1] == "C" && input[i, 2] == "Z") {
            points <- points + 3
       }
       if (input[i, 1] == "A" && input[i, 2] == "Y") {
            points <- points + 6
       }
       if (input[i, 1] == "B" && input[i, 2] == "Z") {
            points <- points + 6
       }
       if (input[i, 1] == "C" && input[i, 2] == "X") {
           points <- points + 6
       }
       if (input[i, 2] == "X") {
           points <- points + 1
       }
       if (input[i, 2] == "Y") {
           points <- points + 2
       }
       if (input[i, 2] == "Z") {
           points <- points + 3
       }
    }
    return(points)
}

rps(input)

### Part 2
input2 <- input
input[, 3] <- NA

for (i in 1:nrow(input2)) {
    if (input2[i, 1] == "A" && input2[i, 2] == "X") {
        input2[i, 3] <- "Z"
    }
    if (input2[i, 1] == "A" && input2[i, 2] == "Y") {
        input2[i, 3] <- "X"
    }
    if (input2[i, 1] == "A" && input2[i, 2] == "Z") {
        input2[i, 3] <- "Y"
    }
    if (input2[i, 1] == "B" && input2[i, 2] == "X") {
        input2[i, 3] <- "X"
    }
    if (input2[i, 1] == "B" && input2[i, 2] == "Y") {
        input2[i, 3] <- "Y"
    }
    if (input2[i, 1] == "B" && input2[i, 2] == "Z") {
        input2[i, 3] <- "Z"
    }
    if (input2[i, 1] == "C" && input2[i, 2] == "X") {
        input2[i, 3] <- "Y"
    }
    if (input2[i, 1] == "C" && input2[i, 2] == "Y") {
        input2[i, 3] <- "Z"
    }
    if (input2[i, 1] == "C" && input2[i, 2] == "Z") {
        input2[i, 3] <- "X"
    }
}

rps(input2[, c(1, 3)])

File: 2023/solutions/02-solution.py
import re, math

file_name = "/Users/tea/Library/Mobile Documents/com~apple~CloudDocs/coding/advent-of-code/2023/input/02-input.txt"
file_content = open(file_name).readlines()

test = ["Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green",
"Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue",
"Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red",
"Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red",
"Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green"]

### first part
def find_possible_games(input, r, g, b):
    possible = []
    bag = {'r':r, 'g':g, 'b':b}
    for line in input:
        for num, col in re.findall(r'(\d+) (\w)', line):
            success = True
            if bag[col] < int(num):
                success = False
                break
        if success:
            possible.append(int(re.search(r'Game (\d+)', line).group(1)))
    return sum(possible)

print(find_possible_games(file_content, 12, 13, 14))

### second part
def find_fewest_cubes(input):
    power = 0
    for line in input:
        bag = {'r':0, 'g':0, 'b':0}
        for num, col in re.findall(r'(\d+) (\w)', line):
            bag[col] = max(bag[col], int(num))
        power += (math.prod(bag.values()))
    return power

print(find_fewest_cubes(file_content))